### ISSUE TITLE
bug: IonicServerModule throws runtime error in Angular 16 with SSR enabled

### ISSUE THREAD

**Issue by SebastianKohler** [ionic-team/ionic-framework#28077]
### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#creating-an-issue).
- [X] I agree to follow the [Code of Conduct](https://ionicframework.com/code-of-conduct).
- [X] I have searched for [existing issues](https://github.com/ionic-team/ionic-framework/issues) that already report this problem, without success.

### Ionic Framework Version

v7.x

### Current Behavior

I have an [Angular 15 app with SSR enabled which uses Ionic 7 web components](https://github.com/slsfi/digital_edition_web_ng), and it's working fine. However, on [upgrading to Angular 16](https://github.com/slsfi/digital_edition_web_ng/tree/angular-16) (which requires zone.js 0.13.x) the `hydrateDocument` function in `IonicServerModule` throws an error on every page load in the server side terminal.

The error is:

```
Error: NotYetImplemented
    at exports2.nyi (C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:2944961)
    at hydrateApp (C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:424924)
    at hydrateAppClosure (C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:1747612)
    at hydrateFactory (C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:1747719)
    at render (C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:1754100)
    at C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:1768972
    at new ZoneAwarePromise (C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:2509981)
    at hydrateDocument (C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:1768620)
    at C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:4033488
    at C:\Users\sebkoh\Documents\GitHub\angular16-ionic7-ssr\dist\angular16-ionic7-ssr\server\main.js:1:3245560
```

Also, it looks like the Ionic components are not rendered properly on the server. In the html response from the SSR an `<ion-button>` element, for instance, is just like in the html template. There's no shadow-root, inner `<button>` element, slots etc.

This is not just a problem in my app, but a general one if you create a new Angular 16 app, enable SSR in it and add Ionic 7 to it (see steps to reproduce below).

### Expected Behavior

No runtime errors in the server side terminal. Ionic components to render properly on the server.

### Steps to Reproduce

1. Create a new Angular 16 app following the official Angular guide https://angular.io/guide/setup-local. Start by opening a terminal window and running `npm install -g @angular/cli`
2. Run `ng new my-app` to create a new Angular 16 app named "my-app" (or anything). When asked about adding Angular routing answer yes.
3. Navigate to the app folder: `cd my-app`
4. Enable SSR (Angular Universal) following the official guide https://angular.io/guide/universal, i.e. run `ng add @nguniversal/express-engine`
5. Add Ionic 7 to dependencies in `package.json`:
```
"@ionic/angular": "^7.3.1",
"@ionic/angular-server": "^7.3.1",
```
6. Run `npm install`
7. Open `app.module.ts` and import IonicModule, so the file becomes:
```
import { NgModule } from '@angular/core';
import { BrowserModule } from '@angular/platform-browser';
import { IonicModule } from '@ionic/angular';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    IonicModule.forRoot(
      {
        mode: 'md',
      }
    ),
    AppRoutingModule
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
```
8. Open `app.server.module.ts` and import `IonicServerModule`, so the file becomes:
```
import { NgModule } from '@angular/core';
import { ServerModule } from '@angular/platform-server';
import { IonicServerModule } from '@ionic/angular-server';

import { AppModule } from './app.module';
import { AppComponent } from './app.component';

@NgModule({
  imports: [
    AppModule,
    ServerModule,
    IonicServerModule,
  ],
  bootstrap: [AppComponent],
})
export class AppServerModule {}
```
9. Run `npm run build:ssr`
10. Run `npm run serve:ssr`
11. Go to http://localhost:4000/ in a browser
12. The above mentioned "NotYetImplemented" error is shown in the terminal every time the page is loaded

To check how Ionic components are rendered on the server, modify `app.component.html` so it contains, for instance, just `<ion-button>Ionic button</ion-button>`

Build and serve the SSR, open DevTools in Chrome and go to the Network tab. Reload http://localhost:4000/ and select the localhost item in the Network tab. Select the Response tab. You will see that the ion-button is rendered as just `<ion-button>Ionic button</ion-button>` on the server, i.e. incomplete.

### Code Reproduction URL

https://github.com/SebastianKohler/angular16-ionic7-ssr

### Ionic Info

I don't use the Ionic CLI. The Angular CLI output is:

```
Angular CLI: 16.2.0
Node: 18.14.2
Package Manager: npm 9.5.0
OS: win32 x64

Angular: 16.2.2
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, platform-server
... router

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1602.0
@angular-devkit/build-angular   16.2.0
@angular-devkit/core            16.2.0
@angular-devkit/schematics      16.2.0
@angular/cli                    16.2.0
@nguniversal/builders           16.2.0
@nguniversal/express-engine     16.2.0
@schematics/angular             16.2.0
rxjs                            7.8.1
typescript                      5.1.6
zone.js                         0.13.1
```

The `@ionic/angular` and `@ionic/angular-server` version is 7.3.1

### Additional Information

I have recently posted this issue in the Angular framework's issues: https://github.com/angular/angular/issues/51319

However, this seems to be a problem with Ionic rather than Angular -- an incompatibility between IonicServerModule and Angular 16 (or zone.js which was updated for Angular 16).

**Comment by patricsteiner** [comment]
Same problem here... thanks for opening the bug report. Would be great to get some insights on this from the ionic-team 🙏🏻

**Comment by sean-perkins** [comment]
Hello @SebastianKohler thanks for reporting this issue! Apologies it has taken us a little time to get to triaging it. 

I am able to reproduce this issue in Angular 16 and can also confirm that it does not reproduce in Angular 15.

The root cause is that the Angular build is incorrectly removing this code from the final bundle: https://github.com/ionic-team/stencil/blob/2f7fb88dcd75312f658421c4c518fa76292517db/src/hydrate/runner/patch-dom-implementation.ts#L40-L52.

This causes the `baseURI` implementation not to be patched for Stencil and throws an exception from domino as a result.

I can verify this behavior by following these steps:
1. Downgrade your application to Angular 15
2. Run `npm run build:ssr`
3. Open `dist/server/main.js` (formatting the file makes the following steps easier)
4. Search for `function patchDomImplementation(e2, t) {`
5. Verify the following code is in the bundle:
```ts
try {
  e2.baseURI;
} catch {
  Object.defineProperty(e2, "baseURI", {
    get() {
      const t3 = e2.querySelector("base[href]");
      return t3
        ? new URL(t3.getAttribute("href"), r.location.href).href
        : r.location.href;
    },
  });
}
``` 

Repeat the steps for your reproduction using Angular 16.

Observe now that the code is just missing. It should be present above `function getRootNode(e2) {`. 

At this time since this behavior only happens when upgrading Angular's dependencies (or deps related to using Angular 16), I don't think this is an issue within Ionic Framework. I honestly do not know what change Angular's ecosystem is contributing to this. I'll follow-up with the Angular team and see if we can collaborate on the root cause from their side.

**Comment by SebastianKohler** [comment]
No worries, and thank you for further investigating the issue. Hopefully you and the Angular team can find a solution.

**Comment by sean-perkins** [comment]
It looks like we may be able to update our source code in Stencil to avoid the compiler optimization treating our detection for the `baseURI` as dead code. I can take a look at this during my R&D time next week.

**Comment by sean-perkins** [comment]
Can you test with this dev-build and let me know if you experience any issues?

```
npm install @ionic/angular@7.5.1-dev.11697219113.17f86185 @ionic/angular-server@7.5.1-dev.11697219113.17f86185
```

**Comment by SebastianKohler** [comment]
@sean-perkins I'm not receiving any errors anymore and as far as I can tell Ionic components are rendered properly in the SSR response, so the dev-build looks good!

**Comment by mic8** [comment]
Hi, is there any followup for this issue? Cause in angular 16, there is still error like this with Angular Universal Imlementation

`Error: NotYetImplemented`

**Comment by sean-perkins** [comment]
@mic8 did you test with the dev-build posted here: https://github.com/ionic-team/ionic-framework/issues/28077#issuecomment-1761950254?

Github issues that are open have not been resolved/released.

**Comment by jwbeech** [comment]
Hi all, any idea when this will make it into a release? This is blocking our upgrade. I can confirm fixed with the dev build: @ionic/angular@7.5.1-dev.11697219113.17f86185 but this is not ideal.

**Comment by patricsteiner** [comment]
same, we're blocked by this... - would be amazing to have this released sooon 😬

**Comment by SebastianKohler** [comment]
This issue seems to have been resolved in `v7.5.8`, right? I updated to `v7.6.0` in an Angular 17 app and server side rendering of Ionic components is working like in the dev-build.

**Comment by sean-perkins** [comment]
@SebastianKohler correct, thanks for the comment!

This ended up being a Stencil bug that we fixed in v4.8.0 of Stencil and was resolved as a result of this commit: https://github.com/ionic-team/ionic-framework/commit/34417a583510009a6ff9749f65d717ea32db6ed8 that shipped in v7.5.8.
